File: convert_html_to_jsx.py
import os
import re
from html.parser import HTMLParser

class JSXConverter(HTMLParser):
    def __init__(self):
        super().__init__(convert_charrefs=False)
        self.output = []
        self.void_elements = {'area', 'base', 'br', 'col', 'embed', 'hr', 'img', 'input', 'link', 'meta', 'param', 'source', 'track', 'wbr'}

    def camel_case(self, s):
        parts = s.split('-')
        return parts[0] + ''.join(p.capitalize() for p in parts[1:])

    def style_to_dict(self, style_str):
        rules = style_str.split(';')
        style_dict = []
        for rule in rules:
            if not rule.strip(): continue
            parts = rule.split(':', 1)
            if len(parts) == 2:
                key = self.camel_case(parts[0].strip())
                val = parts[1].strip()
                # escape single quotes
                val = val.replace("'", "\\'")
                style_dict.append(f"'{key}': '{val}'")
        return "{" + ", ".join(style_dict) + "}"

    def handle_starttag(self, tag, attrs):
        if tag == 'script':
            self.output.append(f'<{tag}>')
            return

        props = []
        for name, value in attrs:
            if name == 'class': name = 'className'
            elif name == 'for': name = 'htmlFor'
            elif name == 'tabindex': name = 'tabIndex'
            elif name == 'readonly': name = 'readOnly'
            elif name == 'maxlength': name = 'maxLength'
            elif name == 'autoplay': name = 'autoPlay'
            elif name == 'playsinline': name = 'playsInline'
            elif name.startswith('on'):
                name = 'data-' + name

            if name == 'style' and value:
                props.append(f"style={{{self.style_to_dict(value)}}}")
            elif name in ['disabled', 'required', 'readOnly', 'multiple', 'checked', 'muted', 'autoPlay']:
                if value is None or value == name or value == '':
                    if name == 'checked' and tag == 'input':
                        props.append('defaultChecked')
                    else:
                        props.append(name)
                else:
                    props.append(f"{name}='{value.replace(chr(39), chr(92)+chr(39))}'")
            elif value is None:
                props.append(name)
            else:
                if name == 'value' and tag in ['input', 'textarea', 'select']:
                    name = 'defaultValue'
                # escape double quotes
                escaped_val = value.replace('"', '&quot;')
                props.append(f'{name}="{escaped_val}"')

        props_str = " " + " ".join(props) if props else ""
        if tag in self.void_elements:
            self.output.append(f'<{tag}{props_str} />')
        else:
            self.output.append(f'<{tag}{props_str}>')

    def handle_endtag(self, tag):
        if tag not in self.void_elements:
            self.output.append(f'</{tag}>')

    def handle_data(self, data):
        # Escape curly braces for JSX
        escaped = data.replace('{', '&#123;').replace('}', '&#125;')
        self.output.append(escaped)

    def handle_comment(self, data):
        self.output.append(f"{{/* {data} */}}")
        
    def handle_entityref(self, name):
        self.output.append(f"&{name};")
        
    def handle_charref(self, name):
        self.output.append(f"&#{name};")

    def handle_decl(self, decl):
        pass

def to_pascal_case(s):
    words = re.split(r'[-_]', s)
    return ''.join(w.capitalize() for w in words)

def extract_body(html_content):
    body_match = re.search(r'<body[^>]*>(.*?)</body>', html_content, re.IGNORECASE | re.DOTALL)
    if body_match:
        return body_match.group(1)
    return html_content

legacy_dir = os.path.join(os.path.dirname(__file__), 'frontend')
react_pages_dir = os.path.join(os.path.dirname(__file__), 'frontend-react', 'src', 'pages')

os.makedirs(react_pages_dir, exist_ok=True)

for filename in os.listdir(legacy_dir):
    if not filename.endswith('.html'):
        continue
        
    with open(os.path.join(legacy_dir, filename), 'r', encoding='utf-8') as f:
        html_content = f.read()
        
    body_content = extract_body(html_content)
    
    parser = JSXConverter()
    parser.feed(body_content)
    jsx_content = "".join(parser.output)
    
    base_name = os.path.splitext(filename)[0]
    component_name = to_pascal_case(base_name)
    if component_name == 'Index':
        component_name = 'Landing'
        
    # We must wrap it in a fragment because multiple roots might exist
    react_code = f"""import React, {{ useEffect }} from 'react';

const {component_name} = () => {{
    useEffect(() => {{
        // Add original JS logic here if needed
    }}, []);

    return (
        <React.Fragment>
            {jsx_content}
        </React.Fragment>
    );
}};

export default {component_name};
"""
    
    out_path = os.path.join(react_pages_dir, f"{component_name}.jsx")
    with open(out_path, 'w', encoding='utf-8') as f:
        f.write(react_code)
    print(f"Converted {filename} to {component_name}.jsx")

File: convert_html_to_jsx.js
const fs = require('fs');
const path = require('path');
const { JSDOM } = require('jsdom');

const legacyDir = path.join(__dirname, 'frontend-legacy');
const reactPagesDir = path.join(__dirname, 'frontend-react', 'src', 'pages');

if (!fs.existsSync(reactPagesDir)) {
    fs.mkdirSync(reactPagesDir, { recursive: true });
}

function camelCase(str) {
    return str.replace(/-([a-z])/g, (g) => g[1].toUpperCase());
}

function styleStringToObject(styleStr) {
    const obj = {};
    const rules = styleStr.split(';');
    for (const rule of rules) {
        if (!rule.trim()) continue;
        const parts = rule.split(':');
        if (parts.length >= 2) {
            const key = camelCase(parts[0].trim());
            const val = parts.slice(1).join(':').trim();
            obj[key] = val;
        }
    }
    return obj;
}

function nodeToJsx(node) {
    if (node.nodeType === 3) { // Text node
        // Escape braces that are not inside JSX blocks (though there shouldn't be any in raw HTML)
        let text = node.textContent;
        // Basic escaping if needed, but usually fine
        return text.replace(/{/g, '&#123;').replace(/}/g, '&#125;');
    }
    if (node.nodeType === 8) { // Comment
        return `{/* ${node.data} */}`;
    }
    if (node.nodeType === 1) { // Element
        const tag = node.tagName.toLowerCase();
        // Skip script tags
        if (tag === 'script') return '';

        let props = '';
        for (const attr of node.attributes) {
            let name = attr.name;
            let value = attr.value;

            // React specific renames
            if (name === 'class') name = 'className';
            else if (name === 'for') name = 'htmlFor';
            else if (name === 'tabindex') name = 'tabIndex';
            else if (name === 'readonly') name = 'readOnly';
            else if (name === 'maxlength') name = 'maxLength';
            else if (name.startsWith('on')) {
                // inline handlers are tricky in React, store as data attributes or ignore
                name = 'data-' + name;
            }

            // Fix invalid characters in attribute names (e.g. data attributes are fine)
            if (name === 'style') {
                const styleObj = styleStringToObject(value);
                props += ` style={${JSON.stringify(styleObj)}}`;
            } else if (name === 'value' && (tag === 'input' || tag === 'textarea' || tag === 'select')) {
                // React requires defaultValue for uncontrolled components
                props += ` defaultValue={${JSON.stringify(value)}}`;
            } else if (name === 'checked' && tag === 'input') {
                props += ` defaultChecked`;
            } else if (name === 'disabled' || name === 'required' || name === 'readOnly' || name === 'multiple') {
                 if(value === '' || value === name) props += ` ${name}`;
                 else props += ` ${name}="${value}"`;
            } else if (name === 'autoplay') {
                 props += ` autoPlay`;
            } else if (name === 'muted') {
                 props += ` muted`;
            } else if (name === 'playsinline') {
                 props += ` playsInline`;
            } else {
                // Escape quotes in values
                const escapedValue = value.replace(/"/g, '&quot;');
                props += ` ${name}="${escapedValue}"`;
            }
        }

        const voidElements = ['area', 'base', 'br', 'col', 'embed', 'hr', 'img', 'input', 'link', 'meta', 'param', 'source', 'track', 'wbr'];
        
        if (voidElements.includes(tag)) {
            return `<${tag}${props} />`;
        } else {
            let childrenJsx = '';
            for (const child of node.childNodes) {
                childrenJsx += nodeToJsx(child);
            }
            // Some specific tag handling can go here
            if (tag === 'textarea') {
                // In React, textarea value is managed via value/defaultValue, not children
                // But if it has children, we convert it to defaultValue
                const text = childrenJsx.trim();
                if (text && !props.includes('defaultValue')) {
                     props += ` defaultValue="${text.replace(/"/g, '&quot;')}"`;
                }
                return `<textarea${props} />`;
            }
            return `<${tag}${props}>${childrenJsx}</${tag}>`;
        }
    }
    return '';
}

function toPascalCase(str) {
    return str.split(/[-_]/).map(word => word.charAt(0).toUpperCase() + word.slice(1)).join('');
}

const files = fs.readdirSync(legacyDir).filter(f => f.endsWith('.html'));

for (const file of files) {
    const filePath = path.join(legacyDir, file);
    const htmlContent = fs.readFileSync(filePath, 'utf-8');
    
    const dom = new JSDOM(htmlContent);
    const body = dom.window.document.body;
    
    // Extract everything inside body, but for index.html maybe ignore the loader/nav?
    // The user wants EXACT SAME TO SAME.
    
    let jsxContent = '';
    for (const child of body.childNodes) {
        jsxContent += nodeToJsx(child);
    }
    
    // Convert base name
    const baseName = path.basename(file, '.html');
    let componentName = toPascalCase(baseName);
    if (componentName === 'Index') componentName = 'Landing';

    const reactFileContent = `import React, { useEffect } from 'react';

const ${componentName} = () => {
    useEffect(() => {
        // Any original inline scripts or logic can be ported here
    }, []);

    return (
        <React.Fragment>
            ${jsxContent}
        </React.Fragment>
    );
};

export default ${componentName};
`;

    const outPath = path.join(reactPagesDir, `${componentName}.jsx`);
    fs.writeFileSync(outPath, reactFileContent, 'utf-8');
    console.log(`Converted ${file} to ${componentName}.jsx`);
}
